src/ProviderPR/SFA.DAS.ProviderPR/Application/Queries/GetRelationshipByEmail/GetRelationshipByEmailQueryHandler.cs

```diff
@@ -0,0 +1,80 @@
+﻿using MediatR;
+using RestEase;
+using SFA.DAS.ProviderPR.Application.Queries.GetRelationship;
+using SFA.DAS.ProviderPR.Infrastructure;
+using SFA.DAS.SharedOuterApi.Configuration;
+using SFA.DAS.SharedOuterApi.InnerApi.Requests;
+using SFA.DAS.SharedOuterApi.InnerApi.Requests.EmployerAccounts;
+using SFA.DAS.SharedOuterApi.InnerApi.Responses;
+using SFA.DAS.SharedOuterApi.InnerApi.Responses.EmployerAccounts;
+using SFA.DAS.SharedOuterApi.Interfaces;
+using SFA.DAS.SharedOuterApi.Models.ProviderRelationships;
+using System.Net;
+
+namespace SFA.DAS.ProviderPR.Application.Queries.GetRelationshipByEmail;
+public class GetRelationshipByEmailQueryHandler(IAccountsApiClient<AccountsConfiguration> accountsApiClient, IProviderRelationshipsApiRestClient providerRelationshipsApiClient) : IRequestHandler<GetRelationshipByEmailQuery, GetRelationshipByEmailQueryResult>
+{
+    public async Task<GetRelationshipByEmailQueryResult> Handle(GetRelationshipByEmailQuery request, CancellationToken cancellationToken)
+    {
+        var hasUserAccount = false;
+        bool? hasOneEmployerAccount = null;
+        bool? hasOneLegalEntity = null;
+
+        var result = await accountsApiClient.Get<GetUserByEmailResponse>(new GetUserByEmailRequest(request.Email));
+
+        if (result == null)
+        {
+            return new GetRelationshipByEmailQueryResult(hasUserAccount, hasOneEmployerAccount, null, hasOneLegalEntity);
```

Comment:
Would suggest using either non-mandatory properties or class based defaults instead of using method based properties as setters, reduces the number of times the properties hasOneLegalEntity and hasOneEmployerAccount have to be referenced before they are actually required for use (would allow them to be completely removed even).

Would allow GetRelationshipByEmailQueryResult(hasUserAccount, hasOneEmployerAccount, null, hasOneLegalEntity) to be reduced to GetRelationshipByEmailQueryResult() when the result is null.

Reply:
I agree. I would have done it slightly differently as well. Since this is an incremental population of result object and we may not need to populate other bits as we have exit route at each step, I would instantiate the result object right at the top (as this needs to be returned no matter what) and then as I go I would collect information directly into the result object and return it as I hit an exit route. It will require all the properties in the result object to have default values or be nullable. And we can see exactly where the exit conditions are. In my opinion it will read lot better.